source/trainer.py

```diff
@@ -0,0 +1,147 @@
+from transformers import (
+    PreTrainedTokenizerFast,
+    DataCollatorForLanguageModeling,
+    GPT2Config,
+    GPT2LMHeadModel,
+    TrainingArguments,
+    Trainer,
+)
+from tokenizers import Tokenizer
+from tokenizers.models import WordLevel
+from tokenizers.pre_tokenizers import WhitespaceSplit
+from tokenizers.trainers import WordLevelTrainer
+from datetime import datetime
+import wandb
+
+# CONFIG:
+TRAIN_FROM_CHECKPOINT = None  # Example: checkpoint-80000
+EVAL_STEPS = 1000
+PER_DEVICE_TRAIN_BATCH_SIZE = 10
+TRAIN_EPOCHS = 5
+
+formattedtime = datetime.now().strftime("%d-%m__%H-%M-%S")
+wandb.init(project=f"the-jam-machine-{formattedtime}")
+
+try:
+    from google.colab import drive
+
+    drive.mount("/content/gdrive")
+    drive_path = Path("/content/gdrive/MyDrive/the_jam_machine")
+    dataset_path = Path(f"{drive_path}/midi_encoded")
+    model_path = Path(f"{drive_path}/model_{formattedtime}")
+except:
+    dataset_path = "./midi_encoded"
+    model_path = f"./models/model_{formattedtime}"
+
+if not os.path.exists(model_path):
+    os.mkdir(model_path)
+
+tokenizer_path = f"{model_path}/tokenizer.json"
+dataset_path_train = os.path.join(dataset_path, "train")
+dataset_path_valid = os.path.join(dataset_path, "validate")
+
+
+def get_dataset(dataset_path):
+    all_files_paths = [f"{dataset_path}/{file}" for file in os.listdir(dataset_path)]
+    dataset = []
+    for file_path in all_files_paths:
+        with open(file_path) as file:
+            for line in file:
+                dataset.append(line.rstrip("\n"))
+    return dataset
+
+
+dataset_train = get_dataset(dataset_path_train)
+dataset_validate = get_dataset(dataset_path_valid)
+
+
+if TRAIN_FROM_CHECKPOINT is None:
+    tokenizer = Tokenizer(WordLevel(unk_token="[UNK]"))
+    tokenizer.pre_tokenizer = WhitespaceSplit()
+    tokenizer_trainer = WordLevelTrainer(special_tokens=["[UNK]", "[PAD]", "[MASK]"])
+    tokenizer.train_from_iterator(dataset_train, trainer=tokenizer_trainer)
+    tokenizer.save(tokenizer_path)
+tokenizer = PreTrainedTokenizerFast(tokenizer_file=tokenizer_path)
+tokenizer.add_special_tokens({"pad_token": "[PAD]"})
+print("Vocabulary size: ", tokenizer.vocab_size)
+
+
+def tokenize_function(data_to_tokenize):
+    tokenized_data = tokenizer(
+        data_to_tokenize,
+        truncation=True,
+        padding=True,
+        max_length=2048,
+    )
+    return {"input_ids": tokenized_data["input_ids"]}
+
+
+dataset_train_tokenized = list(map(tokenize_function, dataset_train))
+dataset_val_tokenized = list(map(tokenize_function, dataset_validate))
+
+
+assert list(dataset_train_tokenized[0]) == ["input_ids"], list(
+    dataset_train_tokenized[0]
+)
+# Check a few samples
+for i, data in enumerate(dataset_train[:3]):
+    print("----")
+    print(data)
+    print(dataset_train_tokenized[i])
+
+model_config = GPT2Config(
+    vocab_size=tokenizer.vocab_size,
+    pad_token_id=tokenizer.pad_token_id,
+    n_embd=512,
+    n_head=8,
+    n_layer=10,
```

Comment:
maybe n_layer could be smaller, like n_layer = 8 ; that will speed up the training. we should check whether it makes a big difference